cartofacade/carto_facade_queue.go

```diff
@@ -0,0 +1,151 @@
+	workType WorkType
+	inputs   map[InputType]interface{}
+}
+
+// DoWork provides the logic to call the correct cgo functions with the correct input.
+func (w *WorkItem) DoWork(
+	q *Queue,
+) (interface{}, error) {
+	switch w.workType {
+	case Initialize:
+		return New(q.CartoConfig, q.CartoAlgoConfig, *q.CartoLib)
+	case Start:
+		return nil, q.Carto.Start()
+	case Stop:
+		return nil, q.Carto.Stop()
+	case Terminate:
+		return q.Carto.Terminate(), nil
+	case AddSensorReading:
+		reading, ok := w.inputs[Reading].([]byte)
+		if !ok {
+			return nil, errors.New("could not cast inputted byte to byte slice")
+		}
+
+		timestamp, ok := w.inputs[Timestamp].(time.Time)
+		if !ok {
+			return nil, errors.New("could not cast inputted timestamp to times.Time")
+		}
+
+		return nil, q.Carto.AddSensorReading(reading, timestamp)
+	case Position:
+		return q.Carto.GetPosition()
+	case InternalState:
+		return q.Carto.GetInternalState()
+	case PointCloudMap:
+		return q.Carto.GetPointCloudMap()
+	}
+	return nil, fmt.Errorf("no worktype found for: %v", w.workType)
+}
+
+// Queue represents a queue to consume work from and enforce one call into C at a time.
+type Queue struct {
+	WorkChannel     chan WorkItem
+	CartoLib        *CartoLib
+	Carto           Carto
+	CartoConfig     CartoConfig
+	CartoAlgoConfig CartoAlgoConfig
+}
+
+// NewQueue instantiates the Queue struct.
+func NewQueue(cartoLib *CartoLib, cartoCfg CartoConfig, cartoAlgoCfg CartoAlgoConfig) Queue {
+	return Queue{
+		WorkChannel:     make(chan WorkItem),
+		Carto:           Carto{},
+		CartoLib:        cartoLib,
+		CartoConfig:     cartoCfg,
+		CartoAlgoConfig: cartoAlgoCfg,
+	}
+}
+
+// HandleIncomingRequest puts incoming requests on the queue and consumes from queue.
+func (q *Queue) HandleIncomingRequest(ctx context.Context, workType WorkType, inputs map[InputType]interface{}) interface{} {
+	// Question: What is reasonable timeout period?
+	ctx, cancel := context.WithTimeout(ctx, 15*time.Second)
+	defer cancel()
+
+	work := WorkItem{
+		Result:   make(chan interface{}),
+		workType: workType,
+		inputs:   inputs,
+	}
+
+	// wait until work can get put on the queue (and timeout if needed)
+	select {
+	case q.WorkChannel <- work:
+		// wait until result is put on result queue (and timeout if needed)
+		result := <-work.Result
+		return result
+	case <-ctx.Done():
+		return nil
+	}
+}
+
+// StartBackgroundWorker starts the background goroutine that is responsible for putting work
+// onto the queue and consuming from the queue.
+func (q *Queue) StartBackgroundWorker(ctx context.Context, activeBackgroundWorkers *sync.WaitGroup) {
+	activeBackgroundWorkers.Add(1)
+
+	goutils.PanicCapturingGo(func() {
+		defer activeBackgroundWorkers.Done()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case workToDo := <-q.WorkChannel:
+				result, err := workToDo.DoWork(q)
+				if err == nil {
+					workToDo.Result <- result
+				} else {
+					workToDo.Result <- err
+				}
```

Comment:
I just realized, that this will cause an issue if I try to cast an error to the wrong type. Should I create a new struct that can hold result and error and change the type of the result channel on `WorkItem` to be `chan resultType`